src/services/rfqm_service.ts

```diff
@@ -439,6 +450,118 @@ export class RfqmService {
     }
 
+    /**
+     * Submit a transaction, re-submit if necessary, and poll for resolution
+     */
+    public async completeSubmissionLifecycleAsync(
+        orderHash: string,
+        workerAddress: string,
+        callData: string,
+    ): Promise<void> {
+        const submissions: { [transactionHash: string]: RfqmTransactionSubmissionEntity } = {};
+
+        let gasPrice = await this._protocolFeeUtils.getGasPriceEstimationOrThrowAsync();
+
+        const firstSubmission = await this._submitTransactionAsync(orderHash, workerAddress, callData, gasPrice);
+        submissions[firstSubmission.transactionHash!] = firstSubmission;
+
+        let isFinishedMonitoring = false;
+        let shouldTryResubmitting = true;
+        while (!isFinishedMonitoring) {
+            await sleepAsync(TRANSACTION_WATCHER_SLEEP_TIME_MS);
+
+            // check if any tx has been mined
+            const receipts = await Promise.all(
+                Object.keys(submissions).map(async (transactionHash) => {
+                    return {
+                        transactionHash,
+                        response: await this._blockchainUtils.getTransactionReceiptIfExistsAsync(transactionHash),
+                    };
+                }),
+            );
+
+            for (const receipt of receipts) {
+                // if a tx has been mined, stop trying to resubmit
+                if (receipt.response !== undefined) {
+                    shouldTryResubmitting = false;
+                    const currentBlock = await this._blockchainUtils.getCurrentBlockAsync();
+                    if (receipt.response.blockNumber - currentBlock >= BLOCK_FINALITY_THRESHOLD) {
+                        isFinishedMonitoring = true;
+                    }
+                    // update all entities
+                    receipts.map((r) => {
+                        if (r.response !== undefined) {
+                            submissions[r.transactionHash].status =
+                                r.response.status === 1
+                                    ? RfqmTranasctionSubmissionStatus.Successful
+                                    : RfqmTranasctionSubmissionStatus.Reverted;
+                            submissions[r.transactionHash].blockMined = new BigNumber(r.response.blockNumber);
+                            submissions[r.transactionHash].gasUsed = new BigNumber(r.response.gasUsed);
+                        } else {
+                            submissions[r.transactionHash].status = RfqmTranasctionSubmissionStatus.DroppedAndReplaced;
+                        }
+                    });
+                    await this.dbUtils.updateRfqmTransactionSubmissionsAsync(Object.values(submissions));
+                    break;
+                }
+            }
+
+            if (shouldTryResubmitting) {
+                const newGasPrice = await this._protocolFeeUtils.getGasPriceEstimationOrThrowAsync();
+
+                if (gasPrice.lt(newGasPrice)) {
```

Comment:
are we going to be constantly resubmitting? (I assume this statement will always be `true`)

Reply:
Should be otherwise only resubmit if gas price changes by a certain threshold?

Reply:
Hmmm.... in theory, there are cases when we will not resubmit (i.e. gas price stays constant). 

In practice, I think @PirosB3 is correct - we will pretty much only see this if the txn wasn't mined because gas prices have increased. It's possible we send txns a little too aggressively if we don't have a threshold

Reply:
I think we should think of submission as free tho (within reason). This is pretty aggressive resubmission but might as well be aggressive to get a better chance at mining